File: core/world/manager.js
import { world } from "./world.js";
import EnemyManager from "../player/ships/enemies/manager.js";
import { keybinds } from "../events/keybind.js";
import { ship } from "../player/ships/player.js";
import WeaponManager from "../weapons/manager.js";

class WorldManager {
    constructor() {
        this.attachedId = -1;
        this.#addKeybinds();
    }

    #addKeybinds() {
        keybinds["5"] = () => {
            this.attachNext();
        }
        keybinds["6"] = () => {
            this.attachPrevious();
        }
        keybinds["7"] = () => {
            this.attachMainShip();
        }
        keybinds["8"] = () => {
            this.attachRandom();
        }

        keybinds["="] = () => {
            world.zoom(1.1);
        }
        keybinds["-"] = () => {
            world.zoom(0.9);
        }
    }


    findAttachedShip() {
        if (this.attachedId === -1) {
            return ship
        } else {
            return EnemyManager.ships[this.attachedId];
        }
    }

    attachRandom() {
        this.attachedId = Math.floor(Math.random() * EnemyManager.ships.length);
        world.attach(EnemyManager.ships[this.attachedId]);

        if (typeof pendo !== "undefined") {
            pendo.track("camera_view_changed", {
                attachment_method: "random",
                target_ship_name: EnemyManager.ships[this.attachedId]?.name,
                attached_ship_index: this.attachedId
            });
        }
    }

    attachNext() {
        this.attachedId = (this.attachedId + 1) % EnemyManager.ships.length;
        world.attach(EnemyManager.ships[this.attachedId]);

        if (typeof pendo !== "undefined") {
            pendo.track("camera_view_changed", {
                attachment_method: "next",
                target_ship_name: EnemyManager.ships[this.attachedId]?.name,
                attached_ship_index: this.attachedId
            });
        }
    }

    attachPrevious() {
        this.attachedId = (this.attachedId - 1 + EnemyManager.ships.length) % EnemyManager.ships.length;
        world.attach(EnemyManager.ships[this.attachedId]);

        if (typeof pendo !== "undefined") {
            pendo.track("camera_view_changed", {
                attachment_method: "previous",
                target_ship_name: EnemyManager.ships[this.attachedId]?.name,
                attached_ship_index: this.attachedId
            });
        }
    }

    attachMainShip() {
        this.attachedId = -1;
        world.attach(ship);

        if (typeof pendo !== "undefined") {
            pendo.track("camera_view_changed", {
                attachment_method: "main_ship",
                target_ship_name: "Player",
                attached_ship_index: -1
            });
        }
    }
}

export const worldManager = new WorldManager();

File: void_front/game/core/world/manager.py
from __future__ import annotations

from game.core.world.world import world
from game.core.events.keybind import keybinds


class WorldManager:
    def __init__(self):
        self.attachedId = -1
        self._add_keybinds()

    def _add_keybinds(self):
        keybinds["5"] = self.attachNext
        keybinds["6"] = self.attachPrevious
        keybinds["7"] = self.attachMainShip
        keybinds["8"] = self.attachRandom
        keybinds["="] = lambda: world.zoom(1.1)
        keybinds["-"] = lambda: world.zoom(0.9)

    def findAttachedShip(self):
        if self.attachedId == -1:
            from game.core.player.ships import player

            return player.ship

        from game.core.player.ships.enemies.manager import EnemyManager

        if 0 <= self.attachedId < len(EnemyManager.ships):
            return EnemyManager.ships[self.attachedId]
        return None

    def attachRandom(self):
        from game.core.player.ships.enemies.manager import EnemyManager

        if not EnemyManager.ships:
            return
        import random

        self.attachedId = random.randrange(len(EnemyManager.ships))
        world.attach(EnemyManager.ships[self.attachedId])

    def attachNext(self):
        from game.core.player.ships.enemies.manager import EnemyManager

        if not EnemyManager.ships:
            return
        self.attachedId = (self.attachedId + 1) % len(EnemyManager.ships)
        world.attach(EnemyManager.ships[self.attachedId])

    def attachPrevious(self):
        from game.core.player.ships.enemies.manager import EnemyManager

        if not EnemyManager.ships:
            return
        self.attachedId = (self.attachedId - 1 + len(EnemyManager.ships)) % len(EnemyManager.ships)
        world.attach(EnemyManager.ships[self.attachedId])

    def attachMainShip(self):
        self.attachedId = -1
        from game.core.player.ships.player import ship

        world.attach(ship)


worldManager = WorldManager()
